Point-in-cell regrid scheme

dkillick opened the pull request:
Added point-in-cell as a regrid scheme. Note that the underlying regridder (`regrid_weighted_curvilinear_to_rectilinear`) still lives in `iris.experimental`, as do the code changes to provide point-in-cell as a regridder.

bjlittle commented on lib/iris/analysis/_interpolation.py:
```diff
@@ -125,7 +125,7 @@ def extend_circular_data(data, coord_dim):
 
 
-def get_xy_dim_coords(cube):
+def get_xy_dim_coords(cube, dim_coords=True):
```
@dkillick I can see why you've done this, to get the 2d auxiliary spatial coordinates for a curvilinear grid, but now the function name is somewhat confused since it can now return non-dimensional coordinates ...

bjlittle commented on lib/iris/experimental/regrid.py:
```diff
@@ -1015,3 +1011,123 @@ def _regrid_indices(cells, depth, points):
+            raise TypeError("'src' must be a Cube")
+
+        src_grid = snapshot_grid(self._src_cube, dim_coords=False)
+        if get_xy_dim_coords(src, dim_coords=False) != src_grid:
```
@dkillick you could avoid using the `snapshot_grid` (hence back-out those changes) and `get_xy_dim_coords` and make the assumption that your expecting x2 2d spatial coordinates in the src, and use a convenience something along the lines of:

```python
    # Get the source cube x and y 2D auxiliary coordinates.
    sx, sy = src_cube.coord(axis='x'), src_cube.coord(axis='y')
```

dkillick commented:
I've reverted the changes to `get_xy_dim_coords` and the associated `snapshot_grid` so that these once more only return dim coords along the `x` and `y` axes. To fill the gap in functionality this has opened back up I've added a static method to the `CurvilinearRegridder` class to locate coordinates along the `x` and `y` axes of the input cubes.

I've also removed some unnecessary imports from the `Test_CurvilinearRegridder` class.

bjlittle commented:
@dkillick Great, thanks :+1:
